vcd/resource_vcd_nsxt_firewall.go

```diff
@@ -0,0 +1,328 @@
+
+	_ = d.Set("edge_gateway_id", edgeGatewayId)
+
+	result := make([]interface{}, len(fwRules))
+
+	for index, value := range fwRules {
+		sourceSlice := extractIdsFromOpenApiReferences(value.SourceFirewallGroups)
+		sourceInterface := convertToTypeSet(sourceSlice)
+		sourceSet := schema.NewSet(schema.HashSchema(&schema.Schema{Type: schema.TypeString}), sourceInterface)
+
+		destinationSlice := extractIdsFromOpenApiReferences(value.DestinationFirewallGroups)
+		destinationInterface := convertToTypeSet(destinationSlice)
+		destinationSet := schema.NewSet(schema.HashSchema(&schema.Schema{Type: schema.TypeString}), destinationInterface)
+
+		appPortProfileSlice := extractIdsFromOpenApiReferences(value.ApplicationPortProfiles)
+		appPortProfileInterface := convertToTypeSet(appPortProfileSlice)
+		appPortProfileSet := schema.NewSet(schema.HashSchema(&schema.Schema{Type: schema.TypeString}), appPortProfileInterface)
+
+		result[index] = map[string]interface{}{
+			"id":           value.ID,
+			"name":         value.Name,
+			"action":       value.Action,
+			"enabled":      value.Enabled,
+			"ip_protocol":  value.IpProtocol,
+			"direction":    value.Direction,
+			"logging":      value.Logging,
+			"sources":      sourceSet,
+			"destinations": destinationSet,
+			"applications": appPortProfileSet,
+		}
+
+	}
+
+	return d.Set("rule", result)
+}
+
+func getNsxtFirewallTypes(d *schema.ResourceData) []*types.NsxtFirewallRule {
+	ruleInterfaceSlice := d.Get("rule").([]interface{})
+	if len(ruleInterfaceSlice) > 0 {
+
+		result := make([]*types.NsxtFirewallRule, len(ruleInterfaceSlice))
+		for index, oneRule := range ruleInterfaceSlice {
+			oneRuleMapInterface := oneRule.(map[string]interface{})
+
+			result[index] = &types.NsxtFirewallRule{
+				Name:       oneRuleMapInterface["name"].(string),
+				Action:     oneRuleMapInterface["action"].(string),
+				Enabled:    oneRuleMapInterface["enabled"].(bool),
+				IpProtocol: oneRuleMapInterface["ip_protocol"].(string),
+				Logging:    oneRuleMapInterface["logging"].(bool),
+				Direction:  oneRuleMapInterface["direction"].(string),
+				Version:    nil,
+			}
+
+			if oneRuleMapInterface["sources"] != nil {
+				sourceGroups := convertSchemaSetToSliceOfStrings(oneRuleMapInterface["sources"].(*schema.Set))
+				result[index].SourceFirewallGroups = convertSliceOfStringsToOpenApiReferenceIds(sourceGroups)
+			}
+
+			if oneRuleMapInterface["destinations"] != nil {
+				sourceGroups := convertSchemaSetToSliceOfStrings(oneRuleMapInterface["destinations"].(*schema.Set))
+				result[index].DestinationFirewallGroups = convertSliceOfStringsToOpenApiReferenceIds(sourceGroups)
+			}
+
+			if oneRuleMapInterface["applications"] != nil {
+				sourceGroups := convertSchemaSetToSliceOfStrings(oneRuleMapInterface["applications"].(*schema.Set))
+				result[index].ApplicationPortProfiles = convertSliceOfStringsToOpenApiReferenceIds(sourceGroups)
+			}
+
```

Comment:
Hm, looks like one more extra line. gofmt missing?

Reply:
Nope. `gofmt` wouldn't complain about single line space. We have it in github tests. This is something that I leave as a separator while developing, but in fact leave it out when it is not needed anymore.. My bad.